### DISCUSSION TITLE
Will JobRunr 8 support Redis at all?

### DISCUSSION THREAD

**Discussion by d-wire** [jobrunr/jobrunr#1158] [General]
My company is currently evaluating JobRunr for use in our Spring Webflux application. However, we're currently using r2dbc for our database driver and it didn't look like(?) JobRunr supported that out of the box. So our next hope was to leverage Redis for our persistence layer. I saw this in the docs though when looking at how to implement Redis:

Note: the LettuceRedisStorageProvider and JedisRedisStorageProvider are deprecated and will be removed from JobRunr 8.

Does this mean there will be a different Redis driver that should be used for JobRunr 8, or does this mean that support for Redis will be removed entirely?

Thanks in advance!

**Comment by rdehuyss** [comment]
Nope, I'm afraid not. Support for Redis will be dropped. 

The effort for supporting Redis is too much (some JobRunr features do not match well with a key-value store) and as no JobRunr Pro customer is using Redis this makes it unsustainable.

**Reply by d-wire** [reply]
Good to know! Is there currently any support for reactive database drivers like r2dbc? Or at least a way to make JobRunr play nice with r2dbc?

**Reply by rdehuyss** [reply]
No - I'm afraid not.

Personally, I've seen a project almost fail due to the complexity Reactive programming adds (a deadlock because one member of the team did not fully grasp the concept). But again - that's my opinion.

As JobRunr now support Virtual Threads out of the box, I also think the need for reactive programming is less. Just running JobRunr on Java 21 will make it automatically switch to Virtual Threads and double the worker count compared to running on old platform threads (so doubling throughput). We tested with various settings (10x the worker count / 100x the worker count) but that resulted in less throughput due to DB connection congestion and memory issues (more GC happening, ...). In our tests, doubling the worker count was the sweet spot resulting in the highest throughput gains.

As it stands - we don't plan to add support for reactive drivers.

In our tests, JobRunr gets easily to 4k jobs / sec on [this hardware](https://www.hetzner.com/dedicated-rootserver/ex44/configurator/#/) and JobRunr Pro v8 gets to 10k jobs / sec thanks to a new smart queueing implementation.

I hope this helps!
